Documentation for refactored SignTool

JohnTortugo opened the pull request:
Closes #387

JohnTortugo commented on src/Microsoft.DotNet.SignTool/README.md:
````diff
@@ -1,66 +1,105 @@
 
-The command line for SignTool is the following:
+```xml
+<PropertyGroup>
````
I've changed it to ItemGroup already.

weshaggard commented on src/Microsoft.DotNet.SignTool/README.md:
```diff
@@ -1,66 +1,105 @@
-- `-intermediateOutputPath`: Defaults to `<outputPath>\Obj`.  
+**Strong Name:** name of the key to use when strong naming the binary. This can be `null` for values which do not require strong name signing such as VSIX files. 
 
+**Target Framework:** valid values include (but are not limited to): `.NETStandard,Version=v2.0`,  `.NETFramework,Version=v4.6.1`, `.NET Core,Version=v2.0`, etc.
```
I added a note in the PR as well but I'm not sure I understand why Target framework is interesting in the signing configuration.

JohnTortugo replied:
I added a comment there as well.

natemcmaster replied:
I had the same question. I'm trying to figure out how the target framework is relevant. Where did you add a comment? I don't see an explanation about why this is needed.

JohnTortugo replied:
My comment is here: https://github.com/dotnet/arcade/pull/410#discussion_r211435459 Which just points to a @tmat comment here: https://github.com/dotnet/arcade/issues/396#issuecomment-410097453 on his words: "The only reason why we need to check TargetFrameworkAttribute is when 2 binaries that have the same strong name (which includes PKT) need to use a different Authenticode certificate for some reason."

tmat replied:
In some cases we need to use different certs for different target frameworks. E.g. https://github.com/dotnet/symreader/blob/master/eng/SignToolData.json

natemcmaster replied:
I prefer the way this was expressed in the SignToolData.json format, i.e. the certs <-> file mapping is done via file path.

weshaggard commented on src/Microsoft.DotNet.SignTool/README.md:
```diff
@@ -1,66 +1,105 @@
+## Valid of Argument Values
 
-The only required argument is `outputPath` and `-config`.  The rest will be inferred to reasonable defaults. Detailed breakdown:
+**Certificate Name:** name of the Authenticode certificate to use for signing.  Valid values include `Microsoft402`, `WindowsPhone623`, `MicrosoftSHA1Win8WinBlue` and `VsixSHA2`.  
```
Is that the full cert list? I suspect there are potnetially others.

JohnTortugo replied:
No, there are others. I'll patch this when I've more.

weshaggard commented on src/Microsoft.DotNet.SignTool/README.md:
```diff
@@ -1,66 +1,105 @@
-- `-msbuildPath`: Path to the MSBuild.exe binary used to run the actual signing process on Microbuild.  The default is to use MSBuid 14.0 standard installation.
-- `-nugetPackagesPath`: Defaults to `~\.nuget\packages`.  Needed to specify the `<Import>` statements for Microbuild.
-- `-intermediateOutputPath`: Defaults to `<outputPath>\Obj`.  
+**Strong Name:** name of the key to use when strong naming the binary. This can be `null` for values which do not require strong name signing such as VSIX files. 
```
It would also be interesting to specify the common strong name values.

JohnTortugo replied:
Got it. Will do that.

weshaggard commented on src/Microsoft.DotNet.SignTool/README.md:
```diff
@@ -1,66 +1,105 @@
+    LogDir="$(ArtifactsLogDir)"
+    PublishUrl="$(PublishUrl)"
+    OrchestrationManifestPath="$(OrchestrationManifestPath)"
+    MSBuildPath="$(DesktopMSBuildPath)"
```
Requiring the MSBuildPath seems redundant given you are already running in msbuild. Is this still a hard requirement?

JohnTortugo replied:
Created issue #490

tmat replied:
This is the path to the desktop msbuild. It's only passed in when we build using Core CLR.

jaredpar commented on src/Microsoft.DotNet.SignTool/SignToolData.md:
````diff
@@ -1,145 +0,0 @@
-
-Here is a sample:
-
-```json
````
The tool still supports this as an inptu mechanism correct (even though it's not the default)?

weshaggard commented on src/Microsoft.DotNet.SignTool/README.md:
```diff
@@ -1,66 +1,105 @@
+    PublishUrl="$(PublishUrl)"
+    OrchestrationManifestPath="$(OrchestrationManifestPath)"
+    MSBuildPath="$(DesktopMSBuildPath)"
+    MicroBuildCorePath="$(NuGetPackageRoot)microbuild.core\$(MicroBuildCoreVersion)"/>
```
Is this required? I only ask because somethings we disable installing the signing plugin on some builds which means this would fail. Also I think we want to be able to run signing verficiation without requiring access to microbuild to do real signing. I think this should only be required if we are going to do signing.

maririos replied:
> Also I think we want to be able to run signing verficiation without requiring access to microbuild to do real signing. 

Agree

>I think this should only be required if we are going to do signing.

Agree

Besides from setting it as `[Required]` doesn't seem like it is actually required unless we are doing real sign.
Unless I'm missing something, it would be good to remove that

JohnTortugo replied:
Created issue #491

tmat replied:
I don't think this tool should be used to do validation that the artifacts are actually signed. That's better done by tools like MicroBuild Codesign Verification VSTS build task, Binskim, etc.

maririos replied:
For verification, I understood running the SignTool and making sure it can find all the files we need to sign, but not really sign them. Like using `DryRun`.

@weshaggard is this what you were thinking here?

weshaggard replied:
Yes I was under the impression we would use this same task in some fashion, I'd perfer we not use DryRun=true for that but instead some other better named parameter for validation.

@tmat were you imagining we use some other mechanism to do the validation? 

>If it's about DryRun then the fact that the microbuild.core package is not installed doesn't matter, since the targets from that package will not be actually invoked.

Agreed that was my exact point.

tmat replied:
> were you imagining we use some other mechanism to do the validation?

Yes, I don't think the tool should be validating that official build actually signed the binaries. We need to do much more validation on official builds than that (APIScan, binskim, etc.) and signing validation should be part of that. No need to duplicate it in this build task.

DryRun is useful for validating the infrastructure around signing all the way until the actual signing.

weshaggard replied:
OK. @maririos do we have an issue tracking the signing validation work then?

maririos commented on src/Microsoft.DotNet.SignTool/README.md:
```diff
@@ -1,66 +1,105 @@
+This is a MSBuild custom task that provides batch signing and verification for MicroBuild environments. The tool is intended to be run as a post-build step and is able to automatically infer the files that need to be signed given a list of container files (.nupkg, .vsix, etc) as input. The high level features of this tool are:
 
-- Performance: The tool operates as a post-build step and uses the minimum number of requests possible.  This can have a dramatic performance improvement over the typical implementation which signs as a post-compile step.  For example, in Roslyn it took build + sign times down from 1-2 hours to 8 minutes. 
-- Verification: Using the `-test` argument the tool can be run as part of a CI leg to verify the consistency of the configuration file.  This enables developers to catch many build and packaging errors that normally would cause a signed build to break post check-in.
```
We should leave the verification section

maririos commented on src/Microsoft.DotNet.SignTool/README.md:
```diff
@@ -1,66 +1,105 @@
-The exclude section is only relevant when VSIX values are being signed.  It's not uncommon to include files in a VSIX which are not built by the repo producing the VSIX.  Such files should not be included in signing (responsibility of the repo that produced them).  
+| Name                      | Type      | Required | Description                                                  |
+| ------------------------- | --------- | -------- | ------------------------------------------------------------ |
+| DryRun                    | bool      | No       | When true the list of files to be signed will be created but won't be sent to the signing server. |
```
for all the ones that are not required, could you please add the default behavior?

maririos commented on src/Microsoft.DotNet.SignTool/README.md:
```diff
@@ -1,66 +1,105 @@
+| StrongNameSignInfo        | Array     | **Yes**  | Should store the default certificate name and strong name to be used for a given Public Key Token. See details below. |
+| FileSignInfo              | Array     | No       | Used to override the default certificate information for specific files and target frameworks combinations. See details below. |
+| MicroBuildCorePath        | Dir Path  | **Yes**  | Path to MicroBuild.Core package directory.                   |
+| MSBuildPath               | Exe path  | !DryRun  | Path to the MSBuild.exe binary used to run the actual signing process on MicroBuild. |
```
instead of actual, consider using `real sign`. Overall the type of signing is `test` or `real` so that might help clarifying

maririos commented on src/Microsoft.DotNet.SignTool/README.md:
```diff
@@ -1,66 +1,105 @@
-- certificate: name of the Authenticode certificate to use for signing.  Valid values include Microsoft402, WindowsPhone623, MicrosoftSHA1Win8WinBlue and VsixSHA2.  More are likely available.  This value must be specified.
-- strongName: name of the key to use when strong naming the binary.  This can be `null` for values which do not require strong name signing such as VSIX files. 
-- values: array of paths, relative to the binaries directory, which will be signed in this manner.  These paths can include `*` globbing for directory names (helps support localization). 
+`MSBuildPath` and `LogDir` are only required if `DryRun == false`.
```
It would be better to have this distinction at the end of the table, after reading what the arguments are.
You could do something like:

**Note:** `MSBuildPath` and `LogDir` are only required if `DryRun == false`.

JohnTortugo replied:
Will change.

natemcmaster commented:
Design question: is there any reason why separate item groups for signing info and file paths is required? I'm trying to figure out why I need two or three item groups to represent a file and its certificate. I was expecting to be able to do something like this:
```xml
<ItemGroup>
  <ItemsToSign Include="file.dll" CertificateName="Microsoft402" />
</ItemGroup>
```

tmat commented:
@natemcmaster We don't want to have to explicitly list all files that needs to be signed and their certificates. Instead we specify default certificates based on PKT and optionally target framework. Only when files need to be signed by a non-default certificate we use an explicit list.

natemcmaster commented:
That's great, and I think you can achieve that without restricting my ability to write a simple file -> cert mapping.

```xml
<ItemGroup>
  <ItemsToSign Include="custom/*.dll" CertificateName="CustomCert" />
  <ItemsToSign Include="default/*.dll"  />
</ItemGroup>
```

tmat commented:
I don't see how it is restricting. 

The full paths of files to be signed are specified in `ItemsToSign`. If any of them have non-default certificate you specify it in `FileSignInfo`.

We need two different item groups since `FileSignInfo` applies on files that are included in containers listed in `ItemsToSign`, while the files themselves are not listed in `ItemsToSign`.

We could recognize `CertificateName` and `StrongName` attributes on `ItemsToSign`, but that just makes it more complicated -- two ways how to specify sign info instead of one. I'd like to understand why the current design is not sufficient first.

JohnTortugo commented:
I think I've addressed all your comments so far regarding the documentation or created new issues for SignTool changes. About the list of certificates / strong names I'll add it as part of the work on this issue: #272 . 

Are there any more changes needed in this document?

natemcmaster commented:
Chatted in person. We can move my question into a different issue. I don't want to block updating docs. :shipit:

maririos commented:
@natemcmaster and @tmat  Issue https://github.com/dotnet/arcade/issues/506 created to keep the discussion there
